k2/csrc/array.h

```diff
@@ -1,69 +1,133 @@
 // k2/csrc/array.h
 
 // Copyright (c)  2020  Xiaomi Corporation (author: Daniel Povey)
+//                      Fangjun Kuang (csukuangfj@gmail.com)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_ARRAY_H_
 #define K2_CSRC_ARRAY_H_
 
-#include <functional>
-#include <limits>
-#include <memory>
+#include <type_traits>
 #include <vector>
 
+#include "glog/logging.h"
+
 namespace k2 {
 
 /*
-   We will use e.g. StridedPtr<int32_t, T> when the stride is not 1, and
+   We will use e.g. StridedPtr<T*, I> when the stride is not 1, and
    otherwise just T* (which presumably be faster).
 */
-template <typename I, typename T>
+template <typename Ptr, typename I = int32_t>
 struct StridedPtr {
-  T *data;
-  I stride;
-  T &operator[](I i) { return data[i]; }
+  using T = typename std::remove_pointer<Ptr>::type;
+
+  T *data;   // it is NOT owned here
+  I stride;  // in number of elements, NOT number of bytes.
+
+  operator T *() { return data; }
+  operator const T *() const { return data; }
+
+  // ++it
+  StridedPtr &operator++() {
+    data += stride;
+    return *this;
+  }
+
+  // it++
+  StridedPtr operator++(int) {
+    StridedPtr tmp(*this);
+    data += stride;
+    return tmp;
+  }
+
+  bool operator==(const StridedPtr &other) const { return data == other.data; }
```

Comment:
if two StridedPtrs are euqal, semantically there will be `data == other.data && stride = other.stride`

Reply:
yeah but we shouldn't ever be comparing them if they didn't come from the same one... I think this is fine for now, but we can put a warning comment.